### PR TITLE
feat(cli): add OpenAPI mock server workflow

### PR DESCRIPTION
**Pull request by odinr**
**Why is this change needed?**

A Fusion application often needs to keep moving before a backend is available, reproduce a backend edge case, or run browser tests without depending on remote services and Entra ID.

Until now, each application had to assemble that workflow itself: alter service discovery in `dev-server.config.ts`, add proxy middleware, maintain separate mock data, and arrange test authentication. Those pieces could drift apart, and a setup that worked locally was not necessarily reproducible in Playwright or CI.

This PR turns that application-specific wiring into one framework workflow. A developer describes a service once from its OpenAPI schema, starts a standalone mock server, and points the application at it. The same definition drives local development, built-app previews, and browser tests.

**What is the current behavior?**

Mocking is spread across several concerns owned by the application:

```text
fetch remote discovery
modify discovered services in dev-server config
register custom proxy routes
implement mock responses in middleware
configure test authentication separately
teach Playwright how to start and stop every custom process
```

This works, but it makes the application responsible for infrastructure behavior and makes mock scenarios difficult to share or validate consistently.

**What is the new behavior?**

A mock service is an executable OpenAPI-backed definition. It owns its schema, deterministic data, route overrides, and discovery behavior in one place.

```text
define service:
  key = "inventory"
  schema = inventory.openapi.json
  discovery mode = new | replace | merge | hidden
  routes:
    GET /items -> deterministic response

start mock backend:
  ffc mock-server ./mocks --port 4010

start application:
  ffc app dev --mock http://localhost:4010

framework behavior:
  read discovery from mock server
  create local proxy routes
  enable opt-in MSAL mock mode
  route application HTTP clients to deterministic mock responses
```

```mermaid
flowchart LR
    A[OpenAPI schema + service module] --> B[ffc mock-server]
    B --> C[Mock service discovery]
    B --> D[Generated and overridden responses]
    E[ffc app dev or serve --mock] --> C
    E --> F[Development proxy]
    G[Browser application] --> F
    F --> D
    H[MSAL mock identity] --> G
```

The workflow supports four service lifecycles:

- `new`: model a service that has not entered discovery yet.
- `replace`: provide a complete local definition for an existing key.
- `merge`: inherit a schema from a selected preset or earlier local layer and override selected behavior. Remote discovery alone is not a merge source for the standalone server.
- `false`: serve an app-owned endpoint without advertising it through discovery.

Without `--mock`, normal `ffc app dev` can still combine real discovery with discovery-visible local definitions. With `--mock`, the application is isolated from remote discovery and uses only the standalone mock server.

**What is the intended behavior or invariant?**

Mocking remains explicit and opt-in:

```text
if command uses --mock:
  discovery = mock server discovery
  authentication = mock MSAL
else:
  discovery = configured remote discovery + eligible local overlays
  authentication = normal application authentication
```

The contracts reviewers should protect are:

- Application code continues to use normal Fusion HTTP clients; it does not know whether a service is real or mocked.
- Service discovery remains the source used to create proxy routes.
- Middleware has final authority over a response; runtime overrides can replace declarative/generated responses and can be reset.
- `FUSION_SPA_MSAL_MOCK` is enabled only by boolean `true` or the exact string `"true"`.
- An optional CLI plugin falls back only when that plugin package is absent. Broken plugin exports or missing transitive dependencies fail visibly.
- Localhost mock proxying works on Linux runners without requiring `<service>.localhost` DNS resolution.
- Starting, stopping, CORS preflight, IPv4, and IPv6 behavior remain valid server lifecycle operations.

**Does this PR introduce a breaking change?**

No. The mock-server packages, `ffc mock-server`, `--mock`, and SPA mock authentication are additive and opt-in. Existing `ffc app dev --env` behavior is preserved and mapped to the runtime environment options.

**Impact assessment:**

- Breaking changes: No.
- Version bump: Minor for the new mock-server packages and additive CLI/SPA APIs, with accompanying dev-server, cookbook, and documentation changesets.
- Consumer impact: Applications can adopt the workflow incrementally. Existing non-mock development remains available.
- Downstream impact: The optional CLI plugin must be installed to use `ffc mock-server`; the CLI gives an installation hint when it is absent.

**Review guidance:**

The fastest end-to-end review is the cookbook:

```sh
pnpm --filter @equinor/fusion-framework-cookbook-app-react-mock-playwright test
```

Conceptually, the Playwright validation does this:

```text
start ffc mock-server
build the cookbook application
start ffc app serve --mock

assert discovery contains:
  people       # preset service with a local override
  aurora-api   # new pre-production service

assert discovery excludes:
  my-api       # direct-only service

call each service through the application's normal HTTP client
assert deterministic responses

set a runtime operation override
assert the next request uses the override
reset mock state
assert the original deterministic response returns
```

```mermaid
sequenceDiagram
    participant Test as Playwright
    participant App as Built Fusion app
    participant Proxy as Dev-server proxy
    participant Mock as OpenAPI mock server

    Test->>Mock: Start standalone server
    Test->>App: Start app with --mock
    App->>Mock: Fetch isolated discovery
    Test->>App: Open service scenario
    App->>Proxy: Request service operation
    Proxy->>Mock: Forward by service key
    Mock-->>App: Deterministic response
    Test->>Mock: Set runtime override
    Test->>App: Repeat operation
    Mock-->>App: Overridden response
    Test->>Mock: Reset
    Test->>App: Repeat operation
    Mock-->>App: Original response
```

For a manual review:

```sh
cd cookbooks/app-react-mock-playwright
pnpm dev:mock
```

Then verify the direct-only, People override, and Aurora pre-production scenarios in the browser. Also review definition precedence, reset behavior, optional-plugin error handling, and the distinction between normal discovery overlays and isolated `--mock` mode.

Validation completed during development includes mock-server unit tests, CLI and SPA tests/typechecks, dev-server proxy regressions, cookbook Playwright tests, Biome, Fusion lint, package TypeScript builds, and the VuePress production build.

**Additional context**

The documentation includes a migration guide from `dev-server.config.ts` mocks, standalone mock-server reference material, Playwright guidance, and the complete cookbook. Consumer-facing package and documentation changes have changesets.

### Checklist

- [x] Confirm completion of the [self-review checklist](https://github.com/equinor/fusion-framework/blob/main/contributing/self-review.md)
- [x] Confirm TSDoc captures intent for functions, hooks, components, classes, and named arrow functions
- [x] Confirm iterator blocks, decision gates, RxJS chains, and complex decisions explain why they exist
- [x] Confirm React logic and derived values are resolved before markup when applicable
- [x] Confirm README/docs are updated for user-facing changes
- [x] Confirm changes to target branch validation
  - _Included files validated_
  - _No new linting warnings_
  - _Not a duplicate PR ([check existing](https://github.com/equinor/fusion-framework/pulls))_
- [x] Confirm adherence to [code of conduct](https://github.com/equinor/fusion-framework/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: packages/utils/openapi-mock-server/src/discovery/load-service-overrides.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  if (extension === '.yml' || extension === '.yaml') {
+    const path = isAbsolute(source) ? source : join(options.baseDir ?? process.cwd(), source);
+    const parsed: unknown = parseYaml(await readFile(path, 'utf-8'));
+    return assertServiceOverrides(parsed, path);
+  }
+
+  // .json, .ts, .js, .mjs (or no extension) resolve through `importConfig`'s own extension probing.
+  const basename = source.replace(/\.(json|ts|mjs|js)$/i, '');
+  const { config } = await importConfig<ServiceOverrides>(basename, { baseDir: options.baseDir });
+  return config;
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/utils/openapi-mock-server/src/discovery/load-service-overrides.ts]: Fixed in 5e955fb0d: `loadServiceOverrides` now runs `assertServiceOverrides` on the config returned by `importConfig` (JSON/script sidecars) before returning it, matching the YAML path's validation.

File: packages/utils/openapi-mock-server/src/server/handle-control-request.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      sendJson(res, 404, { error: `No mocked service registered for "${serviceKey}"` });
+      return;
+    }
+    const body = (await readJsonBody(req)) as MockOverride | undefined;
+    // The override body must carry the "mock" field the schema requires.
+    if (!body || !('mock' in body)) {
+      sendJson(res, 400, { error: 'Expected a JSON body with a "mock" field.' });
+      return;
+    }
+    handle.override(serviceKey, operationId, body);
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/utils/openapi-mock-server/src/server/handle-control-request.ts]: Fixed in 2ac9e2477: the override body is now checked to be a plain object (not an array/primitive) carrying `mock`, and, if present, `status` must be an integer, before `override()` is called.

File: packages/utils/openapi-mock-server/src/server/create-mock-server.ts [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    },
+
+    preset(source) {
+      return handle.use(Object.values(source));
+    },
+
+    async start(options = {}) {
+      await ensureResolved();
+      const server = createServer(requestListener);
+      httpServer = server;
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/utils/openapi-mock-server/src/server/create-mock-server.ts]: Fixed in d3b52671e: `start()` now throws if `httpServer` is already set, instead of silently creating a second server that `close()` couldn't reach.

File: packages/cli/src/cli/plugins/resolve-optional-plugin.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * @throws The original error, for any failure other than the package not being found.
+ */
+export async function resolveOptionalPlugin(
+  packageName: string,
+): Promise<((program: Command) => void) | null> {
+  try {
+    const mod: { default?: (defaults?: unknown) => (program: Command) => void } = await import(
+      packageName
+    );
+    return typeof mod.default === 'function' ? mod.default() : null;
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/cli/src/cli/plugins/resolve-optional-plugin.ts]: Fixed in 3a0684f4d: `resolveOptionalPlugin` now throws a descriptive error when the package resolves but its `default` export isn't a factory function, instead of falling back to the install-hint path.

File: packages/utils/openapi-mock-server/src/presets/fusion/people.fakers.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  return `data:image/svg+xml,${encodeURIComponent(svg)}`;
+}
+
+/** Field-level fakers for `./people.openapi.json` that can't be expressed as a `faker` path string. */
+export default {
+  // Picks its own colors rather than reusing avatarColor's faked value: field fakers are
+  // resolved independently, with no access to a sibling field's generated result.
+  'SuggestionAccount.avatarUrl': fakeAvatar,
+} satisfies FieldFakerMap;
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/utils/openapi-mock-server/src/presets/fusion/people.fakers.ts:14]: Fixed in 5962a873c: `fusionPreset` now imports the shared `FieldFakerMap` default export from `people.fakers.ts` and passes it straight to `withFields()`, so the `fakeAvatar` implementation only exists in this one file.

File: cookbooks/app-react-mock-playwright/src/routes/layout.tsx [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import { Outlet } from '@equinor/fusion-framework-react-router';
+
+/**
+ * Shared page shell for all routes — the router renders matched child routes into the `Outlet`.
+ */
+export default function Layout() {
```
[the remark this answers is not in the document]
Comment (odinr) [reply cookbooks/app-react-mock-playwright/src/routes/layout.tsx]: Fixed in 86199bd42: added an explicit `ReactElement` return type to this component.

File: cookbooks/app-react-mock-playwright/src/routes/index.tsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+export const handle = {
+  route: {
+    description: 'Renders a greeting and a person name, both served by the mock server.',
+  },
+} as const satisfies RouterHandle;
+
+/**
+ * Index route — renders the mocked greeting and person name at the layout's root path.
+ */
+export default function Index() {
```
[the remark this answers is not in the document]
Comment (odinr) [reply cookbooks/app-react-mock-playwright/src/routes/index.tsx]: Fixed in 86199bd42: added an explicit `ReactElement` return type to this component.

File: cookbooks/app-react-mock-playwright/src/components/PersonName.tsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+interface Person {
+  name: string;
+}
+
+/**
+ * Fetches and renders a person's name from the bundled `people` service, overridden by
+ * `mocks/people.overrides.ts`.
+ */
+export const PersonName = () => {
```
[the remark this answers is not in the document]
Comment (odinr) [reply cookbooks/app-react-mock-playwright/src/components/PersonName.tsx]: Fixed in 86199bd42: added an explicit `ReactElement` return type to this component.

File: cookbooks/app-react-mock-playwright/src/components/Greeting.tsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+interface GreetingResponse {
+  message: string;
+}
+
+/**
+ * Fetches and renders a greeting from the `my-api` service discovered through the mock server
+ * (see `mocks/my-api.openapi.json`).
+ */
+export const Greeting = () => {
```
[the remark this answers is not in the document]
Comment (odinr) [reply cookbooks/app-react-mock-playwright/src/components/Greeting.tsx]: Fixed in 86199bd42: added an explicit `ReactElement` return type to this component.

**Comment by Noggling** [comment]
# Concern: mocking an `endpoints`-registered service isn't documented

`endpoints` (`app.config.ts`) is the other supported way to register an HTTP client, alongside
service discovery ([priority table](https://github.com/equinor/fusion-framework/blob/feat/cli-mock-server-plugin/packages/app/docs/http-clients.md)).
`ffc app dev --mock` only covers service discovery, not all app teams use it. There's no equivalent for `endpoints`, and
nothing in the docs says so.

The obvious thing to try, straight from the
[`endpoints` docs](https://github.com/equinor/fusion-framework/blob/feat/cli-mock-server-plugin/packages/cli/docs/application.md#application-config):

```ts
export default defineAppConfig(() => ({
  endpoints: { test: { url: 'http://localhost:4010/test', scopes: [] } },
}));
```

**Fails with a CORS error.** The app runs on `localhost:3000`; the mock server sends no
`Access-Control-Allow-Origin` header, so the browser blocks the cross-origin request outright.

Making it actually work requires all of the following, none of it documented:

1. Hand-write a `dev-server.config.ts` proxy route — a file/mechanism that exists
   ([docs](https://github.com/equinor/fusion-framework/blob/feat/cli-mock-server-plugin/packages/cli/docs/dev-server-config.md))
   but is never linked from the `endpoints` docs.
2. Know the API plugin mounts routes at `/@fusion-api` and strips that prefix before matching
   ([`plugin.ts`](https://github.com/equinor/fusion-framework/blob/feat/cli-mock-server-plugin/packages/vite-plugins/api-service/src/plugin.ts)) —
   so the endpoint's `url` and the route's `match` must agree on it, in two separate files, with
   a silent 404 if you get it wrong.
3. Know that a `proxy`-only route (no `middleware` key) merges differently than the documented
   middleware examples, via a `deepmerge` that branches on `'middleware' in entry`
   ([`merge-dev-server-config.ts`](https://github.com/equinor/fusion-framework/blob/feat/cli-mock-server-plugin/packages/cli/src/lib/merge-dev-server-config.ts)).
4. Restart `ffc app dev` manually after any `app.config.ts` change — unlike `dev-server.config.ts`,
   it isn't watched
   ([`create-dev-server.ts`](https://github.com/equinor/fusion-framework/blob/feat/cli-mock-server-plugin/packages/cli/src/bin/utils/create-dev-server.ts)),
   so a stale server silently keeps the old value.

Even knowing all four, the route was still easy to get subtly wrong (missing slash, unstripped
prefix, drifted `url`/`match` pair) — always a silent 404, never a pointer to the mismatch.

One more wrinkle: manual browsing and Playwright tests need *different* fixes. A test can dodge
CORS entirely with `page.route()` + `route.fetch()` (runs in Node, no browser CORS involved), but
that only fixes the test — a teammate opening the app in a real browser still hits the CORS error
above. A green Playwright suite doesn't mean manual dev works, or vice versa.

## Ask

1. Document the recipe next to `endpoints` in `application.md`/`http-clients.md`, cross-linked to
   `dev-server-config.md`, covering the `/@fusion-api` prefix and the restart caveat.
2. Add a cookbook (or a spec in this one) demonstrating an `endpoints`-registered service mocked
   end-to-end, the way `my-api`/`people` already demonstrate service discovery.

Longer-term: consider giving `endpoints` the same automatic `--mock` treatment service discovery
gets. Bigger conversation, not blocking this PR.

File: packages/utils/openapi-mock-server/src/server/create-mock-server.ts [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    preset(source) {
+      return handle.use(Object.values(source));
+    },
+
+    async start(options = {}) {
+      // A second start() would create an orphaned second HTTP server that close() could never reach.
+      if (httpServer) {
+        throw new Error('start() was already called on this MockServerHandle; call close() first.');
+      }
+      await ensureResolved();
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/utils/openapi-mock-server/src/server/create-mock-server.ts]: Addressed in 86ed2cf0f. `start()` now reserves its in-flight state before source resolution and clears it on completion or failure; a regression test verifies concurrent starts are rejected.

File: packages/utils/openapi-mock-server/src/server/handle-request.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+  seed?: number,
+): Promise<void> {
+  const url = new URL(req.url ?? '/', 'http://localhost');
+  const method = (req.method ?? 'GET').toUpperCase();
+  // Direct-only services run on a different localhost origin than the browser app.
+  res.setHeader('access-control-allow-origin', '*');
+  res.setHeader('access-control-allow-methods', 'GET, HEAD, POST, PUT, PATCH, DELETE, OPTIONS');
+  res.setHeader('access-control-allow-headers', '*');
+  // Preflight only negotiates access; it must not invoke a mock operation.
+  if (method === 'OPTIONS') {
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/utils/openapi-mock-server/src/server/handle-request.ts]: Addressed in 86ed2cf0f. `handle-request.ts` now short-circuits only requests recognized by `isCorsPreflightRequest`; ordinary OpenAPI OPTIONS operations route normally, with coverage for both paths.

File: packages/utils/openapi-mock-server/src/server/handle-control-request.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    }
+    const body = await readJsonBody(req);
+    // The override body must be a plain object carrying "mock", with an integer "status" if present.
+    if (
+      !body ||
+      typeof body !== 'object' ||
+      Array.isArray(body) ||
+      !('mock' in body) ||
+      ('status' in body && !Number.isInteger((body as { status?: unknown }).status))
+    ) {
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/utils/openapi-mock-server/src/server/handle-control-request.ts]: Addressed in 86ed2cf0f. Control-plane overrides now use named payload and HTTP status validators and reject status codes outside 100 through 599 before registration; regressions cover 99 and 600.

File: cookbooks/app-react-mock-playwright/package.json [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  "private": true,
+  "type": "module",
+  "main": "dist/main.js",
+  "scripts": {
+    "build": "ffc app build",
+    "dev": "ffc app dev --env dev",
+    "dev:mock": "concurrently -k 'pnpm mock:server' 'pnpm mock:dev'",
+    "mock:dev": "ffc app dev --env dev --mock http://localhost:4010",
+    "mock:server": "ffc mock-server ./mocks --port 4010",
+    "test": "pnpm test",
```
[the remark this answers is not in the document]
Comment (odinr) [reply cookbooks/app-react-mock-playwright/package.json]: Addressed in 2b9b1c24d. The cookbook `test` script now delegates to `pnpm test:playwright` instead of recursively invoking itself; the cookbook test command completes successfully.

File: vue-press/src/.vuepress/sidebar.ts [hunk 14]
```diff
@@ -266,9 +290,29 @@ export default sidebar({
...
               link: 'dev-server.md',
             },
+            {
+              text: 'Getting started',
+              link: 'dev-server/getting-started.md',
+            },
             {
               text: 'Configuration',
-              link: 'dev-server-config.md',
+              link: 'dev-server/configuration.md',
```
[the remark this answers is not in the document]
Comment (odinr) [reply vue-press/src/.vuepress/sidebar.ts:299]: Addressed in 06c00f227. Both app and portal dev help now point to `/cli/docs/dev-server/configuration.html`, and their TSDoc links use the same published configuration guide.

File: packages/vite-plugins/spa/src/types.ts [hunk 15]
```diff
@@ -119,6 +119,18 @@ export type FusionTemplateEnv = {
      * on initial load.
      */
     requiresAuth: string;
+    /**
+     * When `'true'`, authentication is served by an in-process mock client
+     * instead of Entra ID — no credentials, redirects, or network calls.
+     * Intended for CI/Playwright runs against the dev server.
+     */
+    mock?: string;
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/vite-plugins/spa/src/types.ts]: Addressed in af7d5d35d: `FusionTemplateEnv.msal.mock` now accepts `string | boolean`, matching the runtime toggle contract. The SPA toggle tests and package typecheck pass.

File: packages/utils/openapi-mock-server/src/server/handle-request.ts [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+  req: IncomingMessage,
+  res: ServerResponse,
+  seed?: number,
+): Promise<void> {
+  const url = new URL(req.url ?? '/', 'http://localhost');
+  const method = (req.method ?? 'GET').toUpperCase();
+  // Direct-only services run on a different localhost origin than the browser app.
+  res.setHeader('access-control-allow-origin', '*');
+  res.setHeader('access-control-allow-methods', 'GET, HEAD, POST, PUT, PATCH, DELETE, OPTIONS');
+  res.setHeader('access-control-allow-headers', '*');
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/utils/openapi-mock-server/src/server/handle-request.ts]: Addressed in af7d5d35d: the mock server now echoes `Access-Control-Request-Headers` and falls back to `authorization, content-type`, so authenticated direct requests pass preflight. The focused CORS regression test passes.

File: packages/cli/src/cli/commands/app/dev.command.ts [hunk 17]
```diff
@@ -52,27 +56,35 @@ export const command = createCommand('dev')
...
       '',
-      'See https://equinor.github.io/fusion-framework/cli/docs/dev-server-config.html for configuration options.',
+      'See https://equinor.github.io/fusion-framework/cli/docs/dev-server/configuration.html for configuration options.',
     ].join('\n'),
   )
   .option('--debug', 'Enable debug mode', !!process.env.RUNNER_DEBUG)
   .option('--manifest <path>', 'Path to the app manifest file (app.manifest[.env]?.[ts,js,json])')
   .option('--config <path>', 'Path to the app config file (app.config[.env]?.[ts,js,json])')
-  .option('--env <environment>', 'Runtime environment for the dev server', 'local')
   .option('--port <port>', 'Port for the development server', '3000')
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/cli/src/cli/commands/app/dev.command.ts:70]: Declined intentionally: this command will not expose an `--env` flag, so af7d5d35d does not restore or map it. The app dev command was linted and typechecked after confirming the option is absent.

File: cookbooks/app-react-mock-playwright/src/index.ts [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+]);
+
+/** create a render component */
+const appComponent = createElement(Router, { routes });
+
+/** create React render root component */
+const createApp = (args: ComponentRenderArgs) => makeComponent(appComponent, args, configure);
+
+/** Render function */
+export const renderApp = (el: HTMLElement, args: ComponentRenderArgs): (() => void) => {
```
[the remark this answers is not in the document]
Comment (odinr) [reply cookbooks/app-react-mock-playwright/src/index.ts]: Addressed in af7d5d35d: `renderApp` now documents its mounting intent, host element, render arguments, and returned unmount callback. Cookbook Biome, Fusion lint, and TypeScript checks pass.

File: packages/dev-server/docs/mocking.md [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+
+A preset is a built-in group of service definitions. The default `fusion` preset supplies common
+services that framework modules resolve during startup. Local modules are layered after presets,
+so app-specific behavior can override the baseline.
+
+## Choose a discovery mode
+
+| Mode | Developer scenario |
+| --- | --- |
+| `'merge'` | Override selected behavior of an existing discovered or preset service while inheriting its schema. |
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/dev-server/docs/mocking.md]: Addressed in af7d5d35d: the migration example now supplies a schema and uses `replace`; all `merge` guidance is restricted to selected presets or earlier local layers and states that remote discovery alone is not a merge source. The VuePress production build passes.

**Comment by odinr** [comment]
@Noggling Thanks for digging into this so thoroughly. Your report exposed that the first version only told the service-discovery story and left app-owned `endpoints` users to infer a brittle proxy setup.

A few things have changed since you reviewed it:

- `endpoints` mocking is now documented next to endpoint registration in both `application.md` and `http-clients.md`, including the `app.config` restart caveat.
- The recommended solution is now a direct-only executable service, with `app.config.local.ts` pointing to its `<key>.localhost` origin.
- The mock server now handles browser CORS and requested preflight headers for those per-service origins.
- The cookbook includes this exact `my-api` lifecycle. The app uses `useHttpClient('my-api')`, and Playwright verifies the response through the real browser request rather than `page.route()` interception.
- Localhost proxy handling was made portable for Linux CI, and cookbook Playwright tests are discovered generically by the PR workflow.

Illustratively, the complete setup and validation now look like this:

```ts
// mocks/my-api.mock.ts
export default defineService({
  key: 'my-api',
  serviceDiscovery: false, // App-owned endpoint; do not advertise it in discovery.
  schema: myApiOpenApiSchema,
});

// app.config.local.ts
export default defineAppConfig(() => ({
  endpoints: {
    'my-api': {
      url: 'http://my-api.localhost:4010',
      scopes: [],
    },
  },
}));

// Development lifecycle (illustrative pseudocode)
const mockServer = await run('ffc mock-server');
let appDevServer = await run('ffc app dev');

// app.config files are resolved at startup, so config changes require a restart.
await edit('app.config.local.ts');
appDevServer = await appDevServer.restart();

// The browser uses the normal named Fusion HTTP client.
const client = useHttpClient('my-api');
const response = await client.json('/greeting');

// Browser -> my-api.localhost:4010 -> CORS-enabled mock server -> OpenAPI response
expect(response).toEqual({ message: 'Hello from the mock server!' });

// Playwright exercises the same browser path; no page.route() interception.
await page.goto('/apps/fusion-framework-cookbook-app-react-mock-playwright');
await expect(page.getByTestId('greeting')).toHaveText(
  'Response: Hello from the mock server!',
);
```

One useful correction to the original workaround: this app-owned endpoint no longer needs a handwritten `dev-server.config.ts` route, `/@fusion-api` knowledge, or a duplicated `url`/`match` pair. `/@fusion-api` is the discovery-backed proxy path. A direct-only endpoint remains explicit in `app.config.local.ts`, stays absent from discovery, and is served directly by the mock server.

`--mock` still does not automatically rewrite arbitrary `endpoints`; it switches service discovery to the mock server and enables mock authentication. The endpoint URL remains application-owned configuration. I agree that automatic endpoint treatment is a useful follow-up, but it remains outside this PR.

So both asks are covered now: the endpoint recipe and restart behavior are documented, and the cookbook exercises the same end-to-end path for manual development and Playwright. Thanks again; this made the feature materially clearer and less fragile.

File: packages/cli-plugins/mock-server/src/load-mock-server-config.ts [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+            // Factory configs may derive settings from the supplied base configuration.
+            if (typeof exported === 'function') {
+              return (
+                (await exported({ command: 'serve', mode: 'development', root }, { base })) ?? {}
+              );
+            }
+            return exported ?? {};
+          },
+          // Repeated command runs must not reuse a stale bundled config URL.
+          esBuildOptions: { write: false },
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/cli-plugins/mock-server/src/load-mock-server-config.ts]: Addressed in 665311b27: the loader now keeps `importScript` default file-backed output, so externalized package imports resolve relative to a filesystem module. The regression fixture imports the documented `defineDevServerConfig` package API; all 13 plugin tests and the package typecheck pass.

File: packages/cli-plugins/mock-server/src/load-mock-server-config.ts [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+        baseDir: root,
+        script: {
+          // Config factories receive the same runtime shape as ordinary development serving.
+          resolve: async (module) => {
+            const exported = module.default;
+            // Factory configs may derive settings from the supplied base configuration.
+            if (typeof exported === 'function') {
+              return (
+                (await exported({ command: 'serve', mode: 'development', root }, { base })) ?? {}
+              );
```
[the remark this answers is not in the document]
Comment (odinr) [reply packages/cli-plugins/mock-server/src/load-mock-server-config.ts:52]: Addressed in 665311b27: standalone config factories now receive `environment: "local"` alongside the existing serve/development/root values. The regression derives `mockServer.path` from that value and resolves `local-api-mocks`; all 13 plugin tests and the package typecheck pass.

**Review by Noggling — APPROVED** [review]
